qemu/tests/boot_with_different_vectors.py

```diff
@@ -93,22 +104,101 @@ def check_interrupt(session, vectors):
                 msg += "Command %s got output %s" % (irq_check_cmd, output)
                 test.fail(msg)
         elif vectors == 9 or vectors == 10:
-            if not (re.findall("virtio[0-9]-input", output) and
-                    re.findall("virtio[0-9]-output", output)):
+            if not (re.findall("virtio[0-9]-input", output)
+                    and re.findall("virtio[0-9]-output", output)):
                 msg = "Could not find the virtio device for MSI-X interrupt"
                 msg += " when vectors = %d " % vectors
                 msg += "Command %s got output %s" % (irq_check_cmd, output)
                 test.fail(msg)
 
-    vectors_list = params["vectors_list"]
+    def get_vectors_from_qtree(qtree_info):
+        lines = qtree_info.splitlines()
+        found_target_section = False
+        space_of_section = ''
+        vectors_value = 0
+        for line in lines:
+            if not found_target_section and re.findall(
+                    "dev: virtio-net-pci,", line):
+                found_target_section = True
+                space_of_section = re.findall("^ *", line)
+            if found_target_section:
+                subsection_space = space_of_section[0] + '  '
+                if re.findall('^' + subsection_space + 'vectors = ', line):
+                    result = line.split()
+                    for element in result:
+                        if element.lstrip('-').isdigit():
+                            vectors_value = element
+                    break
+
+        return vectors_value
+
     login_timeout = int(params.get("login_timeout", 360))
-    sub_test = params.get("sub_test_name", "netperf_stress")
-    for vectors in vectors_list.split():
-        vm = boot_guest_with_vectors(vectors)
-        if int(vectors) < 0:
-            continue
+
+    if params.get("wo_vectors", "no") == "yes":
```

Comment:
Seems this condition statement is for two totally different process, it is weird to combine them together. I understand that they call some same functions, but they actually share no steps out of the `if...else` condition. Unlike `with_vhostforce` and `without_vhostforce`, they share most steps with `without_vectors`. So maybe keep the orginal `with_vector` script, just make those functions independent from `run`, and have separate scripts for without_vector, and import the functions from the original script.
New introduced functions `is_nomsi_enalbed` and `get_vectors_from_qtree` might be implemented in the new script, since the original one does not need them.

Reply:
The reason I add new test cases to this file is that the original case is based on some default value like "with_vectors" and no "pci=nomsi" in kernel cmd line. However, these values should be indicated and some original code should be modified according to these value, like "check_msi_support".
I put with/without vectors and with/without vhostforce together on purpose to show the comparison between different parameters.

Reply:
> The reason I add new test cases to this file is that the original case is based on some default value like "with_vectors" and no "pci=nomsi" in kernel cmd line. However, these values should be indicated and some original code should be modified according to these value, like "check_msi_support".

I understand that they share some configurations, so they definitely should be organized within same one .cfg file. And for the functions, so that means the original ones are not compatible enough, they are like some utils functions utilized by more than one test, should be modified to adopt more, and also should be abstract out of the run() function.

> I put with/without vectors and with/without vhostforce together on purpose to show the comparison between different parameters.

Yes, so the .cfg hierachy clearly shows their relations.
Well the scripts are actually two different ones(should be two `run`), that does not make sense to organize them by `if...else` condition.